Mathlib/Data/List/Basic.lean

```diff
@@ -2559,6 +2567,85 @@ theorem foldlRecOn_nil {C : β → Sort*} (op : β → α → β) (b) (hb : C b)
 
+private lemma append_singleton_append_mem_left_of_length_lt {x₁ x₂ z₁ z₂ : List α} {Y₁ Y₂ : α}
+    (together : x₁ ++ [Y₁] ++ z₁ = x₂ ++ [Y₂] ++ z₂) (longer : x₂.length < x₁.length) :
+    Y₂ ∈ x₁ := by
+  have middle := congr_fun (congr_arg get? together) x₂.length
+  rw [
+    append_assoc x₂,
+    get?_append_right x₂.length.le_refl,
+    Nat.sub_self,
+    singleton_append,
+    get?_cons_zero,
+    append_assoc x₁,
+    get?_append longer,
+  ] at middle
+  exact get?_mem middle
+
+private lemma lt_iff_gt_of_add_eq_add
+    {M : Type*} [LinearOrderedAddCommMonoid M] [CovariantClass M M (· + ·) (· < ·)]
+    {a b c d : M} (total : a + b = c + d) :
+    a < c ↔ d < b := by
+  constructor
+  · intro hac
+    by_contra contr
+    rw [not_lt] at contr
+    exact (add_lt_add_of_lt_of_le hac contr).ne total
+  · intro hbd
+    by_contra contr
+    rw [not_lt] at contr
+    rw [add_comm a b, add_comm c d] at total
+    exact (add_lt_add_of_lt_of_le hbd contr).ne total.symm
+
+/- Also works:
+private lemma lt_iff_gt_of_add_eq_add' {M : Type*} [LinearOrderedSemiring M]
+    {a b c d : M} (total : a + b = c + d) :
+    a < c ↔ d < b :=
+  lt_iff_gt_of_add_eq_add total-/
+
+lemma append_singleton_append_inj_of_nmem {x₁ x₂ z₁ z₂ : List α} {Y₁ Y₂ : α}
+    (notin_x : Y₂ ∉ x₁) (notin_z : Y₂ ∉ z₁) :
+    x₁ ++ [Y₁] ++ z₁ = x₂ ++ [Y₂] ++ z₂ ↔ x₁ = x₂ ∧ Y₁ = Y₂ ∧ z₁ = z₂ := by
+  constructor
+  · intro together
+    have xlens : x₁.length = x₂.length
+    · have not_gt : ¬ x₁.length > x₂.length
+      · intro contra_gt
+        apply notin_x
+        exact append_singleton_append_mem_left_of_length_lt together contra_gt
+      have not_lt : ¬ x₁.length < x₂.length
+      · intro contra_lt
+        apply notin_z
+        have reversed := congr_arg reverse together
+        rw [reverse_append_append, reverse_append_append,
+            reverse_singleton, reverse_singleton] at reversed
+        rw [← mem_reverse]
+        apply append_singleton_append_mem_left_of_length_lt reversed
+        rw [length_reverse, length_reverse]
+        have total := congr_arg length together
+        rw [length_append_append, length_append_append, length_singleton, length_singleton] at total
+        rw [← lt_iff_gt_of_add_eq_add total]
+        exact Nat.add_lt_add_right contra_lt 1
+      rw [Nat.not_lt] at not_gt not_lt
+      exact Nat.le_antisymm not_gt not_lt
+    constructor
+    · rw [append_assoc, append_assoc] at together
+      convert congr_arg (take x₁.length) together
+      · rw [take_left]
+      · rw [xlens, take_left]
+    constructor
+    · have chopped := congr_arg (drop x₁.length) together
+      rw [append_assoc, append_assoc, drop_left, xlens, drop_left] at chopped
+      convert congr_arg head? chopped
+      simp
+    · convert congr_arg (drop x₁.length.succ) together
+      · rw [drop_left']
+        rw [length_append, length_singleton]
+      · rw [xlens, drop_left']
+        rw [length_append, length_singleton]
+  · rintro ⟨rfl, rfl, rfl⟩
+    rfl
```

Comment:
I'm intrigued. Is there no easier proof by inducting on `x₁` generalising `x₂` and casing on `x₂`?

Reply:
I am almost sure there exists an easier proof.